### PR TITLE
Add additional artifacts for auto logging

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

- Take over #3369 

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Autologging for scikit-learn logs metric plots (e.g confusion matrix).

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/utils.py [hunk 1]
```diff
@@ -330,44 +441,81 @@ def _get_regressor_metrics(fitted_estimator, fit_args, fit_kwargs):
...
+        )
+        print("4: ", msg)
+        _logger.warning(msg)
+    else:
+        if bool(name_artifact_dict):
+            with TempDir() as tmp:
+                for name, display in name_artifact_dict.items():
+                    filepath = tmp.path("{}.png".format(name))
+                    display.figure_.savefig(filepath)
+                    plt.close(display.figure_)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:516]: Close the figure after saving it to address: https://github.com/mlflow/mlflow/pull/3369#discussion_r484023859.

File: tests/sklearn/test_sklearn_autolog.py [hunk 2]
```diff
@@ -201,42 +201,116 @@ def test_estimator(fit_func_name):
...
+                "{}.png".format("training_precision_recall_curve"),
+            ]
+        )
+
+    assert all(x in artifacts for x in plot_names)
+
+    loaded_model = load_model_by_run_id(run_id)
+    assert_predict_equal(loaded_model, model, X)
+    # verify no figure is open
+    assert len(plt.get_fignums()) == 0
```
Comment (harupy) [inline tests/sklearn/test_sklearn_autolog.py:263]: Assertion to verify no figure is open.

File: mlflow/sklearn/utils.py [hunk 3]
```diff
@@ -129,13 +138,24 @@ def _get_metrics_value_dict(metrics_list):
             _log_warning_for_metrics(metric.name, metric.function, e)
         else:
             metric_value_dict[metric.name] = metric_value
-
     return metric_value_dict
 
 
+def _get_artifacts_list(artifacts_list):
```
Comment (dbczumar) [inline mlflow/sklearn/utils.py:144]: @harupy I wonder if it makes sense to resolve these artifacts one-by-one within the function where we persist the artifacts. Otherwise, we create `n` matplotlib figures here and close them one-by-one as we log them. If we end up logging a larger volume of artifacts eventually / for certain use cases, or if some systems suffer degraded performance as a result of having 10-20 handles open, this could become problematic.

Comment (harupy) [reply mlflow/sklearn/utils.py:144]: Makes sense. I'll make the change.

File: mlflow/sklearn/utils.py [hunk 4]
```diff
@@ -356,18 +469,43 @@ def _log_specialized_estimator_content(fitted_estimator, run_id, fit_args, fit_k
...
+            return
+
+        with TempDir() as tmp_dir:
+            for artifact in artifacts:
+                try:
+                    display = artifact.function(**artifact.arguments)
+                    display.ax_.set_title(artifact.title)
+                    filepath = tmp_dir.path("{}.png".format(artifact.name))
+                    display.figure_.savefig(filepath)
+                    plt.close(display.figure_)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:503]: @dbczumar Resolved artifacts one-by-one (https://github.com/mlflow/mlflow/pull/3423#discussion_r489575277)

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy @willzhan-db !
